packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -71,46 +60,223 @@ class Intent with Diagnosticable {
+/// helps with that, by providing a lifetime for the connection between the
+/// [listener] and the [Action], and by handling the adding and removing of
+/// the [listener] at the right points in the widget lifecycle.
+///
+/// If you listen to an [Action] widget in a widget hierarchy, you should use
+/// this widget. If you are using an [Action] outside of a widget context, then
+/// you must call removeListener yourself.
+@immutable
+class ActionListener extends StatefulWidget {
+  /// Create a const [ActionListener].
+  ///
+  /// The [listener], [action], and [child] arguments must not be null.
+  const ActionListener({
+    Key key,
+    @required this.listener,
+    @required this.action,
+    @required this.child,
+  })  : assert(listener != null),
+        assert(action != null),
+        assert(child != null),
+        super(key: key);
+
+  /// The [ActionListenerCallback] callback to register with the [action].
+  ///
+  /// Must not be null.
+  final ActionListenerCallback listener;
+
+  /// The [Action] that the callback will be registered with.
+  ///
+  /// Must not be null.
+  final Action<Intent> action;
+
+  /// {@macro flutter.widgets.child}
+  final Widget child;
 
   @override
-  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
-    super.debugFillProperties(properties);
-    properties.add(DiagnosticsProperty<LocalKey>('intentKey', intentKey));
+  _ActionListenerState createState() => _ActionListenerState();
+}
+
+class _ActionListenerState extends State<ActionListener> {
+  @override
+  void initState() {
+    super.initState();
+    widget.action.addActionListener(widget.listener);
   }
+
+  @override
+  void didUpdateWidget(ActionListener oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.action == widget.action && oldWidget.listener == widget.listener) {
+      return;
+    }
+    oldWidget.action.removeActionListener(oldWidget.listener);
+    widget.action.addActionListener(widget.listener);
+  }
+
+  @override
+  void dispose() {
+    widget.action.removeActionListener(widget.listener);
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) => widget.child;
+}
+
+/// An abstract [Action] subclass that adds an optional [BuildContext] to the
+/// [invoke] method to be able to provide context to actions.
+///
+/// [ActionDispatcher.invokeAction] checks to see if the action it is invoking
+/// is a [ContextAction], and if it is, supplies it with a context.
+abstract class ContextAction<T extends Intent> extends Action<T> {
+  /// Called when the action is to be performed.
+  ///
+  /// This is called by the [ActionDispatcher] when an action is invoked via
+  /// [Actions.invoke], or when an action is invoked using
+  /// [ActionDispatcher.invokeAction] directly.
+  ///
+  /// This method is only meant to be invoked by an [ActionDispatcher], or by
+  /// its subclasses, and only when [enabled] is true, not to be called
```

Comment:
The "not to be called directly" fragment doesn't seem to fit into the sentence? (feel free to ignore if my feeling of the english language is just wrong here..). Should this maybe just be a separate sentence: "It is not to be called directly.".

Reply:
Yeah, that's redundant.  Removed.